pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -221,6 +214,126 @@ func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeChange},
+	}
+}
+
+// getPersistentVolumeClaimNameFromPod gets pvc names bound to a pod.
+func (pl *VolumeZone) getPersistentVolumeClaimNameFromPod(pod *v1.Pod) []string {
+	var pvcNames []string
+	for i := range pod.Spec.Volumes {
+		volume := pod.Spec.Volumes[i]
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+		pvcName := volume.PersistentVolumeClaim.ClaimName
+		pvcNames = append(pvcNames, pvcName)
+	}
+	return pvcNames
+}
+
+// isSchedulableAfterStorageClassAdded is invoked whenever a StorageClass added. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeZone) isSchedulableAfterStorageClassAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	_, addedStorageClass, err := util.As[*storage.StorageClass](nil, newObj)
+	if err != nil {
+		return framework.QueueAfterBackoff, fmt.Errorf("unexpected objects in isSchedulableAfterStorageClassAdded: %w", err)
+	}
+
+	pvcNames := pl.getPersistentVolumeClaimNameFromPod(pod)
+	for _, pvcName := range pvcNames {
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(pvcName)
+		if s := getErrorAsStatus(err); !s.IsSuccess() {
+			logger.Error(err, "unexpected error during getting PVC", "Namespace", pod.Namespace, "pvc", pvcName)
+			continue
+		}
+		pvName := pvc.Spec.VolumeName
+		if pvName == "" {
+			scName := storagehelpers.GetPersistentVolumeClaimClass(pvc)
+			if scName == addedStorageClass.Name {
+				logger.V(4).Info("StorageClass was created, and matches the pod PVC's StorageClass", "pod", klog.KObj(pod), "storageClass", klog.KObj(addedStorageClass))
+				if (addedStorageClass.VolumeBindingMode != nil) && (*addedStorageClass.VolumeBindingMode == storage.VolumeBindingWaitForFirstConsumer) {
+					logger.V(4).Info("StorageClass's VolumeBindingMode is WaitForFirstConsumer", "storageClass", klog.KObj(addedStorageClass))
+					return framework.QueueAfterBackoff, nil
+				}
+			}
+		}
+	}
+
+	logger.V(4).Info("StorageClass was created, but it doesn't make this pod schedulable", "pod", klog.KObj(pod), "StorageClass", klog.KObj(addedStorageClass))
+	return framework.QueueSkip, nil
+}
+
+// isSchedulableAfterNodeChange is invoked whenever a node added or updated. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeZone) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		return framework.QueueAfterBackoff, fmt.Errorf("unexpected objects in isSchedulableAfterNodeChange: %w", err)
+	}
+
+	podPVTopologies, status := pl.getPVbyPod(context.TODO(), pod)
```

Comment:
Please change `getPVbyPod` receives logger instead of context so that we don't need to have `context.TODO()`